tsfc/fem.py

```diff
@@ -157,421 +169,354 @@ def spanning_degree(element):
-                table, = tables
-            else:
-                table = numpy.array(tables)
-
-            self.tables[(key, cellwise_constant)] = table
+            for c, D, table in tabulator(ufl_element, max_deriv):
+                store[(ufl_element, c, D)].append(table)
 
         if self.integral_type == 'cell':
-            return ein.Literal(table)
+            for key, (table,) in store.iteritems():
+                self.tables[key] = table
         else:
-            f = self.facet[restriction]
+            for key, tables in store.iteritems():
+                table = numpy.array(tables)
+                if len(table.shape) == 2:
+                    # Cellwise constant; must not depend on the facet
+                    assert numpy.allclose(table, table.mean(axis=0, keepdims=True), equal_nan=True)
+                    table = table[0]
+                self.tables[key] = table
 
-            indices = tuple(ein.Index() for i in range(len(table.shape)-1))
-            return ein.ComponentTensor(
-                ein.Indexed(
-                    ein.Literal(table),
-                    (f,) + indices),
-                indices)
+    def __getitem__(self, key):
+        return self.tables[key]
 
 
 class Translator(MultiFunction, ModifiedTerminalMixin, ufl2gem.Mixin):
+    """Contains all the context necessary to translate UFL into GEM."""
 
     def __init__(self, weights, quadrature_index, argument_indices, tabulation_manager,
                  coefficient_map, index_cache):
         MultiFunction.__init__(self)
         ufl2gem.Mixin.__init__(self)
-        self.weights = ein.Literal(weights)
+        integral_type = tabulation_manager.integral_type
+        self.weights = gem.Literal(weights)
         self.quadrature_index = quadrature_index
         self.argument_indices = argument_indices
         self.tabulation_manager = tabulation_manager
-        self.integral_type = tabulation_manager.integral_type
+        self.integral_type = integral_type
         self.coefficient_map = coefficient_map
-        self.cell_orientations = False
-        self.facet = tabulation_manager.facet
         self.index_cache = index_cache
 
-    def get_cell_orientations(self):
-        try:
-            return self._cell_orientations
-        except AttributeError:
-            if self.integral_type.startswith("interior_facet"):
-                result = ein.Variable("cell_orientations", (2, 1))
-            else:
-                result = ein.Variable("cell_orientations", (1, 1))
-            self.cell_orientations = True
-            self._cell_orientations = result
-            return result
+        if integral_type in ['exterior_facet', 'exterior_facet_vert']:
+            self.facet = {None: gem.VariableIndex('facet[0]')}
+        elif integral_type in ['interior_facet', 'interior_facet_vert']:
+            self.facet = {'+': gem.VariableIndex('facet[0]'),
+                          '-': gem.VariableIndex('facet[1]')}
```

Comment:
I agree it's ugly, and I plan to go in that direction. However, I don't plan to use GEM expressions as indices directly. Indices are special, and there are three kinds of them: fixed indices (int), free indices (Index), and variable indices (VariableIndex; they are runtime constant for a single execution of the kernel). I might introduce a common base class (IndexBase) for them. What I plan to do at one point is to make VariableIndex wrap a GEM expression (with no free indices and no shape) instead of string. That is, to use

    gem.VariableIndex(gem.Indexed(gem.Variable("facet"), (0,)))

instead of

    gem.VariableIndex('facet[0]')